### DISCUSSION TITLE
Global coordinate system into local coordinate system

### DISCUSSION THREAD

**Discussion by NordelDev** [dyn4j/dyn4j#188] [Q&A] [answered]
Hi William,

I have a question regarding the transformation of coordinate system.

![grafik](https://user-images.githubusercontent.com/34217758/111041867-b552bc80-843a-11eb-9f90-bc9ad0d6cf13.png)

Basically I want to know the correct way, how I can define a new `Transform` (including translation and rotation) with that I can extract some information about `SimulationBody` and their `BodyFixture`s.

The scenario is the following:
*   I have several `SimulationBody`s in the world (here I use the "global" / standard coordinate system)
*   I want to define a `Transform`, which is then the origin of the "local" coordinate system (new origin and optional rotation)
*   With that "local" coordinate system I want to extract information about the `Transform` of a `SimulationBody`, which also includes adjusted translation & rotation information of the `Shapes` of the attached `BodyFixture`s

Is there an easy way to extract these information?

Right now I do it the following way:

```
// New CoordinateSystemZero
Vector2 coordinateSystemZero = new Vector2(1.0, 2.0);
Transform coordinateSystemZeroTransform = new Transform();
coordinateSystemZeroTransform.setTranslation(coordinateSystemZero);
coordinateSystemZeroTransform.setRotation(new Rotation(Math.toRadians(90.0)));
		
// Transform A (global)
Transform globalTransformA = simulationBodyA.getTransform();
		
// "New" translation based on new CoordinateSystemZero
Vector2 newTranslationA = coordinateSystemZeroTransform.getInverseTransformed(globalTransformA.getTranslation());
```

I don't get, how exactly I can also "translate" the "global" rotation of the `SimulationBody` into the new "local" rotation.

---

Is there a possibility to define simply a `Transform`, which represents the origin and rotation of the new coordinate system and then write a method, which represents the new `Transform` of a `SimulationBody` in the "local" coordinate system, which automatically updates also the `BodyFixtures` (without running the simulation)?

I hope you could understand what I want :D

**Comment by wnbittle** [comment, marked as answer]
I'm may not I fully understand the scenario/question, but let me take a stab at answering.

```java
// define a transform from world coordinates to your custom space
Transform worldToCustom = new Transform();
worldToCustom.setTranslation(1.0, 0.0);
worldToCustom.setRotation(Math.toRadians(30));

// get the body to world coordinate transform
Transform localToWorld = body.getTransform();
// body local point (using the body's local center just as an example, but this could be any local space point)
Vector2 localPoint = body.getLocalCenter();
// convert to world space
Vector2 worldPoint = localToWorld.getTransformed(localCenter);

// convert to the desired space
Vector2 customPoint = worldToCustom.getInverseTransformed(worldCenter);

// let's say you want to convert back
worldPoint = worldToCustom.getTransformed(customPoint);
localPoint = localToWorld.getInverseTransformed(worldPoint);
```

This applies to points.  For direction vectors, it's all the same except using the `getTransformedR` and `getInverseTransformedR` methods.

**Reply by NordelDev** [reply]
Sorry for the late reply. Saw the notification about your reply just now.

Thank you for your example. I think I had a misunderstanding about how a rotation of the "new" coordinate system is taken into consideration. I thought, that I have to do additionally some calculations so that the rotation of the "new" coordinate system is also applied... My mistake sorry :)
